File: backend/api/views/trainings.py
import json
from django.utils import timezone
from django.db.models import Q
from rest_framework import status
from rest_framework.decorators import api_view, permission_classes
from rest_framework.permissions import IsAuthenticated
from rest_framework.response import Response

from api.models import Employee, Training
from api.permissions import IsAdmin, IsAdminOrEncoder
from api.serializers import TrainingListSerializer, TrainingSerializer
from api.utils import calc_expiration, days_from_today, log_audit, parse_training_validity, today


@api_view(['GET'])
@permission_classes([IsAuthenticated])
def training_summary(request):
    current = today()
    in30 = days_from_today(30)
    in10 = days_from_today(10)

    active_qs = Training.objects.filter(is_archived=False)
    total = active_qs.count()
    expired = active_qs.filter(
        expiration_date__isnull=False,
        expiration_date__lt=current,
    ).count()
    expiring30 = active_qs.filter(
        expiration_date__isnull=False,
        expiration_date__gte=current,
        expiration_date__lte=in30,
    ).count()
    expiring60 = active_qs.filter(
        expiration_date__isnull=False,
        expiration_date__gte=current,
        expiration_date__lte=in10,
    ).count()
    total_employees = Employee.objects.filter(status='active').count()

    return Response({
        'total': total,
        'expired': expired,
        'expiring30': expiring30,
        'expiring60': expiring60,
        'totalEmployees': total_employees,
    })


@api_view(['GET'])
@permission_classes([IsAuthenticated])
def training_categories(request):
    categories = list(
        Training.objects.filter(is_archived=False).exclude(category='')
        .values_list('category', flat=True)
        .distinct()
        .order_by('category')
    )
    return Response(categories)


@api_view(['GET'])
@permission_classes([IsAuthenticated])
def training_titles(request):
    titles = list(
        Training.objects.filter(is_archived=False).exclude(title='')
        .values_list('title', flat=True)
        .distinct()
        .order_by('title')
    )
    return Response(titles)


@api_view(['GET', 'POST'])
@permission_classes([IsAuthenticated])
def training_list_create(request):
    if request.method == 'GET':
        qs = Training.objects.select_related('employee').filter(is_archived=False)
        employee_id = request.query_params.get('employee_id')
        category = request.query_params.get('category')
        search = request.query_params.get('search')

        if employee_id:
            qs = qs.filter(employee_id=employee_id)
        if category:
            qs = qs.filter(category=category)
        worker_line_status = request.query_params.get('worker_line_status')
        if worker_line_status:
            qs = qs.filter(worker_line_status=worker_line_status)
        take = request.query_params.get('take')
        if take:
            qs = qs.filter(take=int(take))
        training_date = request.query_params.get('training_date')
        if training_date:
            qs = qs.filter(training_date=training_date)
        date_from = request.query_params.get('date_from')
        date_to = request.query_params.get('date_to')
        if date_from:
            qs = qs.filter(training_date__gte=date_from)
        if date_to:
            qs = qs.filter(training_date__lte=date_to)
        if search:
            qs = qs.filter(
                Q(employee__full_name__icontains=search)
                | Q(employee__employee_id__icontains=search)
                | Q(title__icontains=search)
            )

        current = today()
        in10 = days_from_today(10)
        if request.query_params.get('expired') == 'true':
            qs = qs.filter(expiration_date__isnull=False, expiration_date__lt=current)
        elif request.query_params.get('expiring_soon') == 'true':
            qs = qs.filter(
                expiration_date__isnull=False,
                expiration_date__gte=current,
                expiration_date__lte=in10,
            )

        return Response(
            TrainingListSerializer(qs.order_by('-training_date'), many=True).data
        )

    if not IsAdminOrEncoder().has_permission(request, None):
        return Response({'error': 'Insufficient permissions'}, status=status.HTTP_403_FORBIDDEN)

    employee_id = request.data.get('employee_id')
    title = (request.data.get('title') or '').strip()
    training_date = request.data.get('training_date')

    if not employee_id or not title or not training_date:
        return Response(
            {'error': 'employee_id, title, and training_date are required'},
            status=status.HTTP_400_BAD_REQUEST,
        )

    try:
        employee = Employee.objects.get(pk=employee_id)
    except Employee.DoesNotExist:
        return Response({'error': 'Employee not found'}, status=status.HTTP_404_NOT_FOUND)

    validity_months, validity_days = parse_training_validity(request.data)
    expiration_date = calc_expiration(
        training_date,
        validity_months=validity_months,
        validity_days=validity_days,
    )

    training = Training.objects.create(
        employee=employee,
        title=title,
        category=request.data.get('category') or '',
        training_date=training_date,
        trainer=request.data.get('trainer') or '',
        validity_months=validity_months,
        validity_days=validity_days,
        expiration_date=expiration_date,
        process_classification=request.data.get('process_classification') or '',
        remarks=request.data.get('remarks') or '',
        worker_line_status=request.data.get('worker_line_status') or 'Floating',
        take=request.data.get('take') or 1,
        created_by_id=request.user.id,
    )
    log_audit(request.user, 'CREATE', 'trainings', training.id, f'Created training: {title}')
    return Response(TrainingSerializer(training).data, status=status.HTTP_201_CREATED)


@api_view(['GET', 'PUT', 'DELETE'])
@permission_classes([IsAuthenticated])
def training_detail(request, pk):
    try:
        training = Training.objects.select_related('employee').get(pk=pk)
    except Training.DoesNotExist:
        return Response({'error': 'Training record not found'}, status=status.HTTP_404_NOT_FOUND)

    if request.method == 'GET':
        data = TrainingSerializer(training).data
        data['employee_name'] = training.employee.full_name
        data['emp_code'] = training.employee.employee_id
        return Response(data)

    if request.method == 'PUT':
        if not IsAdminOrEncoder().has_permission(request, None):
            return Response({'error': 'Insufficient permissions'}, status=status.HTTP_403_FORBIDDEN)

        # Capture before state
        before = {
            'title': training.title,
            'category': training.category,
            'training_date': str(training.training_date),
            'trainer': training.trainer,
            'validity_months': (
                float(training.validity_months) if training.validity_months is not None else None
            ),
            'validity_days': training.validity_days,
            'expiration_date': str(training.expiration_date) if training.expiration_date else None,
            'process_classification': training.process_classification,
            'worker_line_status': training.worker_line_status,
            'take': training.take,
            'remarks': training.remarks,
        }

        new_date = request.data.get('training_date') or training.training_date
        validity_months, validity_days = parse_training_validity(
            request.data,
            default_months=training.validity_months if training.validity_months is not None else 12,
        )

        training.title = request.data.get('title') or training.title
        if 'category' in request.data:
            training.category = request.data['category']
        training.training_date = new_date
        if 'trainer' in request.data:
            training.trainer = request.data['trainer']
        training.validity_months = validity_months
        training.validity_days = validity_days
        training.expiration_date = calc_expiration(
            new_date,
            validity_months=validity_months,
            validity_days=validity_days,
        )
        if 'process_classification' in request.data:
            training.process_classification = request.data['process_classification']
        if 'remarks' in request.data:
            training.remarks = request.data['remarks']
        if 'worker_line_status' in request.data:
            training.worker_line_status = request.data['worker_line_status']
        if 'take' in request.data:
            training.take = request.data['take']
        training.save()

        after = {
            'title': training.title,
            'category': training.category,
            'training_date': str(training.training_date),
            'trainer': training.trainer,
            'validity_months': (
                float(training.validity_months) if training.validity_months is not None else None
            ),
            'validity_days': training.validity_days,
            'expiration_date': str(training.expiration_date) if training.expiration_date else None,
            'process_classification': training.process_classification,
            'worker_line_status': training.worker_line_status,
            'take': training.take,
            'remarks': training.remarks,
        }
        # Only keep changed fields
        changes = {k: {'before': before[k], 'after': after[k]} for k in before if before[k] != after[k]}

        log_audit(
            request.user, 'UPDATE', 'trainings', training.id,
            json.dumps({'summary': f'Updated training: {training.title}', 'changes': changes}),
        )
        return Response(TrainingSerializer(training).data)

    if not IsAdmin().has_permission(request, None):
        return Response({'error': 'Insufficient permissions'}, status=status.HTTP_403_FORBIDDEN)

    training.is_archived = True
    training.archived_at = timezone.now()
    training.save()
    log_audit(request.user, 'ARCHIVE', 'trainings', pk, f'Archived training: {training.title}')
    return Response({'message': 'Training record archived'})


@api_view(['GET'])
@permission_classes([IsAuthenticated])
def archived_trainings(request):
    if not IsAdmin().has_permission(request, None):
        return Response({'error': 'Insufficient permissions'}, status=403)
    qs = Training.objects.select_related('employee').filter(is_archived=True).order_by('-archived_at')
    return Response(TrainingListSerializer(qs, many=True).data)


@api_view(['POST'])
@permission_classes([IsAuthenticated])
def restore_training(request, pk):
    if not IsAdmin().has_permission(request, None):
        return Response({'error': 'Insufficient permissions'}, status=403)
    try:
        training = Training.objects.get(pk=pk, is_archived=True)
    except Training.DoesNotExist:
        return Response({'error': 'Archived record not found'}, status=404)
    training.is_archived = False
    training.archived_at = None
    training.save()
    log_audit(request.user, 'RESTORE', 'trainings', pk, f'Restored training: {training.title}')
    return Response({'message': 'Training record restored'})


@api_view(['DELETE'])
@permission_classes([IsAuthenticated])
def delete_archived_training(request, pk):
    if not IsAdmin().has_permission(request, None):
        return Response({'error': 'Insufficient permissions'}, status=403)
    try:
        training = Training.objects.get(pk=pk, is_archived=True)
    except Training.DoesNotExist:
        return Response({'error': 'Archived record not found'}, status=404)
    title = training.title
    training.delete()
    log_audit(request.user, 'DELETE', 'trainings', pk, f'Permanently deleted training: {title}')
    return Response({'message': 'Training record permanently deleted'})


@api_view(['POST'])
@permission_classes([IsAuthenticated])
def bulk_archive_trainings(request):
    if not IsAdmin().has_permission(request, None):
        return Response({'error': 'Insufficient permissions'}, status=403)
    ids = request.data.get('ids', [])
    if not ids or not isinstance(ids, list):
        return Response({'error': 'A list of ids is required'}, status=400)
    from django.utils import timezone
    qs = Training.objects.filter(pk__in=ids, is_archived=False)
    count = qs.count()
    qs.update(is_archived=True, archived_at=timezone.now())
    log_audit(request.user, 'DELETE', 'trainings', None, f'Bulk archived {count} training record(s)')
    return Response({'message': f'{count} record(s) archived', 'count': count})


@api_view(['POST'])
@permission_classes([IsAuthenticated])
def bulk_restore_trainings(request):
    if not IsAdmin().has_permission(request, None):
        return Response({'error': 'Insufficient permissions'}, status=403)
    ids = request.data.get('ids', [])
    if not ids or not isinstance(ids, list):
        return Response({'error': 'A list of ids is required'}, status=400)
    qs = Training.objects.filter(pk__in=ids, is_archived=True)
    count = qs.count()
    qs.update(is_archived=False, archived_at=None)
    log_audit(request.user, 'RESTORE', 'trainings', None, f'Bulk restored {count} training record(s)')
    return Response({'message': f'{count} record(s) restored', 'count': count})


@api_view(['DELETE'])
@permission_classes([IsAuthenticated])
def bulk_delete_trainings(request):
    if not IsAdmin().has_permission(request, None):
        return Response({'error': 'Insufficient permissions'}, status=403)
    ids = request.data.get('ids', [])
    if not ids or not isinstance(ids, list):
        return Response({'error': 'A list of ids is required'}, status=400)
    qs = Training.objects.filter(pk__in=ids, is_archived=True)
    count = qs.count()
    qs.delete()
    log_audit(request.user, 'DELETE', 'trainings', None, f'Bulk permanently deleted {count} training record(s)')
    return Response({'message': f'{count} record(s) permanently deleted', 'count': count})

File: client/electron/db/trainings.js
'use strict';

const { getPool, ensureSchemaReady } = require('./pool');
const { logAudit } = require('./audit');
const { requireAuth, requireRole } = require('./auth');
const { apiError, dt, today, daysFromToday, calcExpiration, parseTrainingValidity, normalizeCertRecert } = require('./helpers');

function normalizePassFail(value) {
  return value === 'Failed' ? 'Failed' : 'Passed';
}

function serializeDetail(row) {
  return {
    id: row.id,
    employee_id: row.employee_id,
    title: row.title,
    category: row.category,
    training_date: row.training_date || null,
    trainer: row.trainer,
    validity_months: row.validity_months,
    validity_days: row.validity_days,
    expiration_date: row.expiration_date || null,
    process_classification: row.process_classification,
    remarks: row.remarks,
    worker_line_status: row.worker_line_status,
    cert_recert: normalizeCertRecert(row.cert_recert),
    pass_fail: normalizePassFail(row.pass_fail),
    take: row.take,
    created_by: row.created_by,
    created_at: dt(row.created_at),
    updated_at: dt(row.updated_at),
  };
}

function serializeList(row) {
  return {
    ...serializeDetail(row),
    is_archived: !!row.is_archived,
    archived_at: dt(row.archived_at),
    employee_name: row.employee_name,
    emp_code: row.emp_code,
    factory: row.factory,
    line: row.line,
    team: row.team,
  };
}

const LIST_SELECT = `SELECT t.*, e.full_name AS employee_name, e.employee_id AS emp_code,
  e.factory AS factory, e.line AS line, e.team AS team
  FROM trainings t JOIN employees e ON e.id = t.employee_id`;

async function findById(id, archived = null) {
  const pool = getPool();
  let sql = `${LIST_SELECT} WHERE t.id = ?`;
  const vals = [id];
  if (archived === true) sql += ' AND t.is_archived = 1';
  if (archived === false) sql += ' AND t.is_archived = 0';
  const [rows] = await pool.query(sql + ' LIMIT 1', vals);
  return rows[0] || null;
}

async function list(params = {}) {
  requireAuth();
  await ensureSchemaReady();
  const pool = getPool();
  const where = ['t.is_archived = 0'];
  const vals = [];

  if (params.employee_id) { where.push('t.employee_id = ?'); vals.push(params.employee_id); }
  if (params.factory) { where.push('e.factory = ?'); vals.push(params.factory); }
  if (params.category) { where.push('t.category = ?'); vals.push(params.category); }
  if (params.title) { where.push('t.title LIKE ?'); vals.push(`%${params.title}%`); }
  if (params.worker_line_status) { where.push('t.worker_line_status = ?'); vals.push(params.worker_line_status); }
  if (params.cert_recert) { where.push('t.cert_recert = ?'); vals.push(normalizeCertRecert(params.cert_recert)); }
  if (params.pass_fail) { where.push('t.pass_fail = ?'); vals.push(normalizePassFail(params.pass_fail)); }
  if (params.take) { where.push('t.take = ?'); vals.push(parseInt(params.take, 10)); }
  if (params.training_date) { where.push('t.training_date = ?'); vals.push(params.training_date); }
  if (params.date_from) { where.push('t.training_date >= ?'); vals.push(params.date_from); }
  if (params.date_to) { where.push('t.training_date <= ?'); vals.push(params.date_to); }
  if (params.search) {
    where.push('(e.full_name LIKE ? OR e.employee_id LIKE ? OR t.title LIKE ?)');
    const like = `%${params.search}%`;
    vals.push(like, like, like);
  }

  const current = today();
  const in10 = daysFromToday(10);
  if (params.expired === 'true') {
    where.push('t.expiration_date IS NOT NULL AND t.expiration_date < ?');
    vals.push(current);
  } else if (params.expiring_soon === 'true') {
    where.push('t.expiration_date IS NOT NULL AND t.expiration_date >= ? AND t.expiration_date <= ?');
    vals.push(current, in10);
  }

  const sql = `${LIST_SELECT} WHERE ${where.join(' AND ')} ORDER BY t.training_date DESC, t.id DESC`;
  const [rows] = await pool.query(sql, vals);
  return rows.map(serializeList);
}

async function listForEmployee(employeeId) {
  const pool = getPool();
  const [rows] = await pool.query(
    'SELECT * FROM trainings WHERE employee_id = ? ORDER BY training_date DESC',
    [employeeId]
  );
  return rows.map(serializeDetail);
}

async function summary() {
  requireAuth();
  const pool = getPool();
  const current = today();
  const in30 = daysFromToday(30);
  const in10 = daysFromToday(10);
  const one = async (sql, vals) => {
    const [r] = await pool.query(sql, vals);
    return Number(r[0].c);
  };
  const total = await one('SELECT COUNT(*) c FROM trainings WHERE is_archived = 0');
  const expired = await one(
    'SELECT COUNT(*) c FROM trainings WHERE is_archived = 0 AND expiration_date IS NOT NULL AND expiration_date < ?', [current]);
  const expiring30 = await one(
    'SELECT COUNT(*) c FROM trainings WHERE is_archived = 0 AND expiration_date IS NOT NULL AND expiration_date >= ? AND expiration_date <= ?', [current, in30]);
  const expiring60 = await one(
    'SELECT COUNT(*) c FROM trainings WHERE is_archived = 0 AND expiration_date IS NOT NULL AND expiration_date >= ? AND expiration_date <= ?', [current, in10]);
  const totalEmployees = await one("SELECT COUNT(*) c FROM employees WHERE status = 'active'");
  return { total, expired, expiring30, expiring60, totalEmployees };
}

async function categories() {
  requireAuth();
  const pool = getPool();
  const [rows] = await pool.query("SELECT DISTINCT category FROM trainings WHERE is_archived = 0 AND category <> '' ORDER BY category");
  return rows.map((r) => r.category);
}

async function titles() {
  requireAuth();
  const pool = getPool();
  const [rows] = await pool.query("SELECT DISTINCT title FROM trainings WHERE is_archived = 0 AND title <> '' ORDER BY title");
  return rows.map((r) => r.title);
}

async function create(data = {}) {
  const user = requireRole('admin', 'encoder');
  await ensureSchemaReady();
  const pool = getPool();

  const employeeId = data.employee_id;
  const title = (data.title || '').trim();
  const trainingDate = data.training_date;
  if (!employeeId || !title || !trainingDate) {
    throw apiError('employee_id, title, and training_date are required', 400);
  }

  const [emp] = await pool.query('SELECT id FROM employees WHERE id = ? LIMIT 1', [employeeId]);
  if (!emp.length) throw apiError('Employee not found', 404);

  const { validity_months, validity_days } = parseTrainingValidity(data);
  const expiration = calcExpiration(trainingDate, validity_months, validity_days);
  const certRecert = normalizeCertRecert(data.cert_recert);
  const passFail = normalizePassFail(data.pass_fail);

  const [res] = await pool.query(
    `INSERT INTO trainings
      (employee_id, title, category, training_date, trainer, validity_months, validity_days,
       expiration_date, process_classification, remarks, worker_line_status, cert_recert, pass_fail, take,
       is_archived, created_by, created_at, updated_at)
     VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, 0, ?, NOW(6), NOW(6))`,
    [
      employeeId, title, data.category || '', trainingDate, data.trainer || '',
      validity_months, validity_days, expiration,
      data.process_classification || '', data.remarks === 'INACTIVE' ? 'INACTIVE' : 'ACTIVE',
      data.worker_line_status || 'Floating',
      certRecert,
      passFail,
      data.take || 1, user.id,
    ]
  );
  const id = res.insertId;
  await logAudit(user, 'CREATE', 'trainings', id, `Created training: ${title}`);
  const row = await findById(id);
  return serializeDetail(row);
}

async function get({ id } = {}) {
  requireAuth();
  const row = await findById(id);
  if (!row) throw apiError('Training record not found', 404);
  const data = serializeDetail(row);
  data.employee_name = row.employee_name;
  data.emp_code = row.emp_code;
  return data;
}

async function update({ id, data } = {}) {
  const user = requireRole('admin', 'encoder');
  await ensureSchemaReady();
  const pool = getPool();
  const training = await findById(id);
  if (!training) throw apiError('Training record not found', 404);
  data = data || {};

  const snapshot = (t) => ({
    title: t.title,
    category: t.category,
    training_date: t.training_date ? String(t.training_date).slice(0, 10) : null,
    trainer: t.trainer,
    validity_months: t.validity_months != null ? Number(t.validity_months) : null,
    validity_days: t.validity_days,
    expiration_date: t.expiration_date ? String(t.expiration_date).slice(0, 10) : null,
    process_classification: t.process_classification,
    worker_line_status: t.worker_line_status,
    cert_recert: normalizeCertRecert(t.cert_recert),
    pass_fail: normalizePassFail(t.pass_fail),
    take: t.take,
    remarks: t.remarks,
  });
  const before = snapshot(training);

  const newDate = data.training_date || (training.training_date ? String(training.training_date).slice(0, 10) : training.training_date);
  const { validity_months, validity_days } = parseTrainingValidity(
    data,
    training.validity_months != null ? Number(training.validity_months) : 12
  );

  const next = {
    title: data.title || training.title,
    category: 'category' in data ? data.category : training.category,
    training_date: newDate,
    trainer: 'trainer' in data ? data.trainer : training.trainer,
    validity_months,
    validity_days,
    expiration_date: calcExpiration(newDate, validity_months, validity_days),
    process_classification: 'process_classification' in data ? data.process_classification : training.process_classification,
    remarks: 'remarks' in data
      ? (data.remarks === 'INACTIVE' ? 'INACTIVE' : 'ACTIVE')
      : (training.remarks === 'INACTIVE' ? 'INACTIVE' : 'ACTIVE'),
    worker_line_status: 'worker_line_status' in data ? data.worker_line_status : training.worker_line_status,
    cert_recert: 'cert_recert' in data
      ? normalizeCertRecert(data.cert_recert)
      : normalizeCertRecert(training.cert_recert),
    pass_fail: 'pass_fail' in data
      ? normalizePassFail(data.pass_fail)
      : normalizePassFail(training.pass_fail),
    take: 'take' in data ? data.take : training.take,
  };

  await pool.query(
    `UPDATE trainings SET
       title = ?, category = ?, training_date = ?, trainer = ?, validity_months = ?, validity_days = ?,
       expiration_date = ?, process_classification = ?, remarks = ?, worker_line_status = ?,
       cert_recert = ?, pass_fail = ?, take = ?,
       updated_at = NOW(6)
     WHERE id = ?`,
    [
      next.title, next.category, next.training_date, next.trainer, next.validity_months, next.validity_days,
      next.expiration_date, next.process_classification, next.remarks, next.worker_line_status,
      next.cert_recert, next.pass_fail, next.take, id,
    ]
  );

  const after = snapshot({ ...next });
  const changes = {};
  for (const key of Object.keys(before)) {
    if (before[key] !== after[key]) changes[key] = { before: before[key], after: after[key] };
  }
  await logAudit(user, 'UPDATE', 'trainings', id, JSON.stringify({ summary: `Updated training: ${next.title}`, changes }));

  const row = await findById(id);
  return serializeDetail(row);
}

// DELETE on a training soft-archives it (matches Django training_detail DELETE).
async function remove({ id } = {}) {
  const user = requireRole('admin');
  const pool = getPool();
  const training = await findById(id);
  if (!training) throw apiError('Training record not found', 404);
  await pool.query('UPDATE trainings SET is_archived = 1, archived_at = NOW(6), updated_at = NOW(6) WHERE id = ?', [id]);
  await logAudit(user, 'ARCHIVE', 'trainings', id, `Archived training: ${training.title}`);
  return { message: 'Training record archived' };
}

async function archived() {
  requireRole('admin');
  const pool = getPool();
  const [rows] = await pool.query(`${LIST_SELECT} WHERE t.is_archived = 1 ORDER BY t.archived_at DESC`);
  return rows.map(serializeList);
}

async function restore({ id } = {}) {
  const user = requireRole('admin');
  const pool = getPool();
  const training = await findById(id, true);
  if (!training) throw apiError('Archived record not found', 404);
  await pool.query('UPDATE trainings SET is_archived = 0, archived_at = NULL, updated_at = NOW(6) WHERE id = ?', [id]);
  await logAudit(user, 'RESTORE', 'trainings', id, `Restored training: ${training.title}`);
  return { message: 'Training record restored' };
}

async function deletePermanent({ id } = {}) {
  const user = requireRole('admin');
  const pool = getPool();
  const training = await findById(id, true);
  if (!training) throw apiError('Archived record not found', 404);
  const title = training.title;
  await pool.query('DELETE FROM trainings WHERE id = ?', [id]);
  await logAudit(user, 'DELETE', 'trainings', id, `Permanently deleted training: ${title}`);
  return { message: 'Training record permanently deleted' };
}

function normalizeIds(ids) {
  if (!Array.isArray(ids) || ids.length === 0) throw apiError('A list of ids is required', 400);
  return ids.map((v) => Number(v)).filter((v) => Number.isFinite(v));
}

async function bulkArchive({ ids } = {}) {
  const user = requireRole('admin');
  const pool = getPool();
  const list_ = normalizeIds(ids);
  const [res] = await pool.query(
    'UPDATE trainings SET is_archived = 1, archived_at = NOW(6), updated_at = NOW(6) WHERE id IN (?) AND is_archived = 0',
    [list_]
  );
  const count = res.affectedRows || 0;
  await logAudit(user, 'DELETE', 'trainings', null, `Bulk archived ${count} training record(s)`);
  return { message: `${count} record(s) archived`, count };
}

async function bulkRestore({ ids } = {}) {
  const user = requireRole('admin');
  const pool = getPool();
  const list_ = normalizeIds(ids);
  const [res] = await pool.query(
    'UPDATE trainings SET is_archived = 0, archived_at = NULL, updated_at = NOW(6) WHERE id IN (?) AND is_archived = 1',
    [list_]
  );
  const count = res.affectedRows || 0;
  await logAudit(user, 'RESTORE', 'trainings', null, `Bulk restored ${count} training record(s)`);
  return { message: `${count} record(s) restored`, count };
}

async function bulkDelete({ ids } = {}) {
  const user = requireRole('admin');
  const pool = getPool();
  const list_ = normalizeIds(ids);
  const [res] = await pool.query('DELETE FROM trainings WHERE id IN (?) AND is_archived = 1', [list_]);
  const count = res.affectedRows || 0;
  await logAudit(user, 'DELETE', 'trainings', null, `Bulk permanently deleted ${count} training record(s)`);
  return { message: `${count} record(s) permanently deleted`, count };
}

module.exports = {
  list, listForEmployee, summary, categories, titles,
  create, get, update, remove, archived, restore, deletePermanent,
  bulkArchive, bulkRestore, bulkDelete,
  serializeDetail, serializeList,
};
